File: server/dashboard/src/lib/server/logger.ts
import { mkdirSync, appendFileSync, existsSync } from "fs";
import { join } from "path";

let logFilePath: string | null = null;

/**
 * Initialize the logger. Creates the logs/ directory if needed
 * and opens the log file in append mode.
 */
export function initLogger(filename = "dashboard.log"): void {
  const logsDir = join(process.cwd(), "logs");
  if (!existsSync(logsDir)) {
    mkdirSync(logsDir, { recursive: true });
  }
  logFilePath = join(logsDir, filename);
  Log("LOGGER", "Logger initialized. Writing to:", logFilePath);
}

/**
 * Log a timestamped, source-tagged message to stdout and the log file.
 * Format: [YYYY-MM-DD] - [HH:MM:SS] - [source] - message
 */
export function Log(source: string, ...args: unknown[]): void {
  const now = new Date();
  const date = now.toISOString().slice(0, 10);
  const time = now.toTimeString().slice(0, 8);
  const msg = args
    .map((a) => (typeof a === "object" ? JSON.stringify(a) : String(a)))
    .join(" ");

  const line = `[${date}] - [${time}] - [${source}] - ${msg}`;

  console.log(line);

  if (logFilePath) {
    try {
      appendFileSync(logFilePath, line + "\n");
    } catch {
      // If file write fails, stdout logging still works
    }
  }
}

File: backend/logger/logger.go
package logger

import (
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

var (
	Logger  = log.New(os.Stdout, "", 0)
	logFile *os.File
)

// InitLogger initializes the logger to write to both stdout and a file in AppData.
func InitLogger() error {
	configDir, err := os.UserConfigDir()
	if err != nil {
		return err
	}

	logsDir := filepath.Join(configDir, "EMLy", "logs")
	if err := os.MkdirAll(logsDir, 0755); err != nil {
		return err
	}

	logPath := filepath.Join(logsDir, "app.log")
	file, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	logFile = file

	multi := io.MultiWriter(os.Stdout, file)
	Logger = log.New(multi, "", 0)

	Log("Logger initialized. Writing to: " + logPath)
	return nil
}

// CloseLogger closes the log file.
func CloseLogger() {
	if logFile != nil {
		logFile.Close()
	}
}

// LogDepth prints a timestamped log line. skip is passed to runtime.Caller to
// determine the source location: 1 = direct caller, 2 = caller's caller, etc.
func LogDepth(skip int, args ...any) {
	now := time.Now()
	date := now.Format("2006-01-02")
	tm := now.Format("15:04:05")

	_, file, line, ok := runtime.Caller(skip)
	loc := "unknown"
	if ok {
		loc = fmt.Sprintf("%s:%d", filepath.Base(file), line)
	}

	msg := fmt.Sprintln(args...)
	Logger.Printf("[%s] - [%s] - [%s] - %s", date, tm, loc, strings.TrimRight(msg, "\n"))
}

// Log prints a timestamped, file:line tagged log line.
func Log(args ...any) {
	LogDepth(2, args...)
}
